rllib/offline/estimators/off_policy_estimator.py

```diff
@@ -31,18 +40,100 @@ def __init__(self, policy: Policy, gamma: float):
+
+    @DeveloperAPI
+    @OverrideToImplementCustomLogic
+    def estimate_single_step(self, batch: SampleBatch, **kwargs) -> Dict[str, Any]:
+        """Returns off-policy estimates for the batch of single timesteps. This is
+        highly optimized for bandits assuming each episode is a single timestep.
+
+        Args:
+            batch: The batch to calculate the off-policy estimates (OPE) on. The
+            batch must be a sample batch type that contains the fields "obs",
+            "actions", and "action_prob".
+
+        Returns:
+            The off-policy estimates (OPE) calculated on the given episode. The returned
+            dict can be any arbitrary mapping of strings to metrics.
+        """
+        raise NotImplementedError
+
+    @DeveloperAPI
+    def estimate(
+        self, batch: SampleBatchType, split_by_episode: bool = True
+    ) -> Dict[str, Any]:
+        """Compute off-policy estimates.
 
         Args:
             batch: The batch to calculate the off-policy estimates (OPE) on. The
             batch must contain the fields "obs", "actions", and "action_prob".
+            split_by_episode: Whether to split the batch by episode.
 
         Returns:
             The off-policy estimates (OPE) calculated on the given batch. The returned
             dict can be any arbitrary mapping of strings to metrics.
+            The dict consists of the following metrics:
+            - v_behavior: The discounted return averaged over episodes in the batch
+            - v_behavior_std: The standard deviation corresponding to v_behavior
+            - v_target: The estimated discounted return for `self.policy`,
+            averaged over episodes in the batch
+            - v_target_std: The standard deviation corresponding to v_target
+            - v_gain: v_target / max(v_behavior, 1e-8)
+            - v_delta: The difference between v_target and v_behavior.
         """
-        raise NotImplementedError
+        batch = self.convert_ma_batch_to_sample_batch(batch)
+        self.check_action_prob_in_batch(batch)
+        estimates_per_epsiode = []
+        if split_by_episode:
+            for episode in batch.split_by_episode():
+                assert len(set(episode[SampleBatch.EPS_ID])) == 1, (
+                    "The episode must contain only one episode id. For some reason "
+                    "the split_by_episode() method could not successfully split "
+                    "the batch by episodes. Each row in the dataset should be "
+                    "one episode. Check your evaluation dataset for errors."
+                )
+                estimate_step_results = self.estimate_multi_step(episode)
+                estimates_per_epsiode.append(estimate_step_results)
+        else:
+            if is_overridden(self.estimate_single_step):
+                estimates_per_epsiode.append(self.estimate_single_step(batch))
+            else:
+                raise NotImplementedError(
```

Comment:
Good point. I don't know what's the best way to do this. I wanted to give more informative error message because they can also go on with multi_step by setting `ope_split_by_episode=True`

Reply:
so why not just raise this specific message in ``estimate_single_step()`` right above?

Reply:
because that function does not need to know about how it's called? like based on `ope_split_by_episode` ?

Reply:
Another way would be to try and catch the error and change the error message.

Reply:
I see. huh ok.
but however we want to do this, should we do the same for self.estimate_multi_step() call above as well?

Reply:
estimate_multi_step() always works. the other one only works on bandits. An estimator down the line may only implement one or the other or both and should raise not implemented error. The error message should tell the user that hey if you unintentionally wanted to do estimate_multi_step() but your split_by_eps is False and estimate_single_step is not implemented. you should make it True. I think of a better way.

Reply:
wait, I think we are mixing some things up here.
if you look at the two API calls above, they are both NotImplemented.
this makes me feel like the knowledge of ``multi-step will always work`` is coming from the children implementations that we already have, and it's not from the perspective of an abstract API class. like, if someone is only reading the ``OffPolicyEstimator`` class by itself, they will get curious why we catch in one place, but not the other.
either, it's fine to have the knowledge leak from children class to the APIs here, in which case, we should be able to write a more specific error message for the APIs, or we keep the API class pure, and makes no assumption about children implementation.
not trying to nit-picking, you get my idea, just trying to tell the story from a reader of our code.
thanks man.

Reply:
oh you are right. I agree with you on the leakage problem. We don't want this info to leak from children to the base parent abstract class. I'll remove the override check then. 

These nit-pickings are actually important imo. Shoot more :)